Karl/test.cc

```diff
@@ -322,6 +466,165 @@ TEST(Karl, Deregister) {
+                       "/",
+                       karl_nginx_base_url,
+                       nginx_parameters);
+  const current::karl::Locator karl_locator(Printf("http://localhost:%d", FLAGS_karl_test_port));
+
+  current::karl::ClaireStatus claire;
+  claire.service = "unittest";
+  claire.codename = "ABCDEF";
+  claire.local_port = PickPortForUnitTest();
+  // Register fake service.
+  HTTP(claire.local_port).Register("/.current", [](Request r) { r("GOTIT\n"); });
+
+  std::string proxied_url;
+  {
+    const std::string keepalive_url = Printf("%s?codename=%s&port=%d",
+                                             karl_locator.address_port_route.c_str(),
+                                             claire.codename.c_str(),
+                                             claire.local_port);
+    const auto response = HTTP(POST(keepalive_url, claire));
+    EXPECT_EQ(200, static_cast<int>(response.code));
+    while (karl.ActiveServicesCount() == 0u) {
+      ;  // Spin lock.
+    }
+    const auto result =
+        karl.InternalExposeStorage()
+            .ReadOnlyTransaction([&](ImmutableFields<unittest_karl_t::storage_t> fields) {
+              ASSERT_TRUE(Exists(fields.claires[claire.codename]));
+              EXPECT_EQ(current::karl::ClaireRegisteredState::Active,
+                        Value(fields.claires[claire.codename]).registered_state);
+              ASSERT_TRUE(Exists(Value(fields.claires[claire.codename]).url_status_page_proxied));
+              proxied_url = Value(Value(fields.claires[claire.codename]).url_status_page_proxied);
+              EXPECT_EQ("http://localhost:" + current::ToString(FLAGS_karl_nginx_port) + "/proxied/" +
+                            claire.codename,
+                        proxied_url);
+            })
+            .Go();
+    EXPECT_TRUE(WasCommitted(result));
+  }
+  {
+    while (true) {
+      try {
+        const auto response = HTTP(GET(proxied_url));
+        if (response.code == HTTPResponseCode.OK) {
+          EXPECT_EQ("GOTIT\n", response.body);
+          break;
+        }
+      } catch (const current::net::NetworkException& e) {
+      }
+    }
+  }
+
+  current::time::SetNow(std::chrono::microseconds(100 * 1000 * 1000),
+                        std::chrono::microseconds(101 * 1000 * 1000));
+  while (karl.ActiveServicesCount() > 0u) {
+    ;  // Spin lock.
+  }
+  {
+    const auto result = karl.InternalExposeStorage()
+                            .ReadOnlyTransaction([&](ImmutableFields<unittest_karl_t::storage_t> fields) {
+                              ASSERT_TRUE(Exists(fields.claires[claire.codename]));
+                              EXPECT_EQ(current::karl::ClaireRegisteredState::DisconnectedByTimeout,
+                                        Value(fields.claires[claire.codename]).registered_state);
+                            })
+                            .Go();
+    EXPECT_TRUE(WasCommitted(result));
+  }
+
+  // This is a bit flaky.
```

Comment:
Why is it? As nginx reloads the config, this route should return 404; why won't it?